### PR TITLE
[8.x] Small correction on Factory::resolveFactoryName for packages

### PR DESCRIPTION
**Pull request by N3XT0R**
Actually on creating custom packages for laravel where the $namespace attribute is set on factory classes
its impossible to let laravel guess the name by Factory::resolveFactoryName correctly.

This is just a small bugfix on it to resolve the names correctly.
An additional test i've added, too. All other tests work fine, too.

### CODE REVIEW

File: tests/Database/DatabaseEloquentFactoryTest.php [hunk 1]
```diff
@@ -441,6 +441,22 @@ public function test_resolve_non_app_nested_model_factories()
...
     }
 
+    public function test_resolve_package_model_factories(): void
+    {
+        Factory::useNamespace('Vendor\\Package\\Database\\Factories\\');
+
+        $resolves = [
+            'Vendor\\Package\\Models\\Foo' => 'Vendor\\Package\\Database\\Factories\\FooFactory',
+            'Vendor\\Package\\Models\\Bar' => 'Vendor\\Package\\Database\\Factories\\BarFactory',
+            'Vendor\\Package\\Models\\Sub\\Foo' => 'Vendor\\Package\\Database\\Factories\\FooFactory',
```
Comment (rodrigopedra) [inline tests/Database/DatabaseEloquentFactoryTest.php:451]: If `Vendor\Package\Models\Sub\Foo` is a different model than `Vendor\Package\Models\Foo`, shouldn't they use different factory classes?

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by N3XT0R** [comment]
@taylorotwell 

i am sure that this is a bug. Just try itself by creating a package with a Model and a Factory let us say:
```php
namespace My\Package\Database\Factories;

class MyTestModelFactory extends Factory
{
    protected $model = MyTestModel::class;

    protected static $namespace = 'My\\Package\\Database\\Factories\\'; //<-- different namespace than default

    public function definition(): array
    {
        return [
            'any_column' => $this->faker->word,
        ];
    }
}
``` 
and one equivalent model to the factory ( i think an explanation for a model is not really useful here). 

when someone installs a package that contains this factory too this person is not able to use factory classes from packages to write own unittests for testing with third party components (future tests).

Because of:
```php
/**
     * Get the factory name for the given model name.
     *
     * @param  string  $modelName
     * @return string
     */
    public static function resolveFactoryName(string $modelName)
    {
        //input of modelName will be My\Package\Models\MyTestModel
        $resolver = static::$factoryNameResolver ?: function (string $modelName) { 
            $appNamespace = static::appNamespace(); // <-- appNamespace will return e.g. "App\"

            $modelName = Str::startsWith($modelName, $appNamespace.'Models\\') //<-- will return false because App\Models\\ is not a part of the given modelName
                ? Str::after($modelName, $appNamespace.'Models\\')
                : Str::after($modelName, $appNamespace); //<-- modelName will be here the full qualified class name of given model

            return static::$namespace.$modelName.'Factory'; //<-- here the static::$namespace 
"My\\Package\\Database\\Factories\\" will be prepended (same as defined in factory class)
        };

        return $resolver($modelName); //result will be App\My\Package\Database\Factories\My\Package\Model\MyTestModelFactory
    }
```

in earlier versions of laravel (before 8) it was possible to use vendor factories too.

File: tests/Database/DatabaseEloquentFactoryTest.php (hunk 1 above)
Comment (N3XT0R) [reply tests/Database/DatabaseEloquentFactoryTest.php:451]: you are right. My Idea on it wasn't to change the actual implementation to much. Would i implement on my pull request a resolving with sub folders, so anyone would need to write an additional entry inside of the documentation for it (i thought).

**Comment by rodrigopedra** [comment]
@N3XT0R on your package provided models you can override the `HasFactory`'s `newFactory()` method to new up a custom factory:

~~~php
<?php

namespace Vendor\Package;

use Vendor\Package\Factories\FavoriteFactory;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Favorite extends Model
{
    use HasFactory;
    
    protected static function newFactory()
    {
        return new FavoriteFactory();
    }
}
~~~

That is an empty method on `HasFactory`'s trait meant to be overridden to provide a factory placed on a custom location:

https://github.com/laravel/framework/blob/bf3d5e319f0bd928a493222adb1acd9d23c1db68/src/Illuminate/Database/Eloquent/Factories/HasFactory.php#L22-L30

**Pull request closed without merging**